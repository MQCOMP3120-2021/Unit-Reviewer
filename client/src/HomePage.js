import React, { useState, useEffect } from 'react'
import { Link } from 'react-router-dom'
import { Grid, Image, Card, Icon, Rating, Pagination } from 'semantic-ui-react'

const HomePage = ({ units, unitsLength, getUnits, activePage, setActivePage }) => {
    console.log(Math.ceil(unitsLength / 10))

    useEffect(() => {
        getUnits(activePage)
    }, [activePage])

    return (<>
        <Grid padded centered>
            <Grid.Row>
                {units.map(item => (
                    <Link key={item._id} to={`unit/${item._id}`}><Card style={{ marginBottom: 10, marginTop: 10, marginRight: 10 }}>
                        <Card.Content header={item.code} />
                        <Card.Content description={item.title} />
                        <Card.Content extra>
                            <Rating icon='star'
                                defaultRating={item.reviews.length > 0 && item.reviews.map(rev => rev.rating).reduce((a, b) => (a + b)) / item.reviews.length}
                                disabled maxRating={5} /> ({item.reviews.length} Ratings)
                        </Card.Content>
                    </Card></Link>
                ))}
            </Grid.Row>
            <Grid.Row>
                <Pagination
                    onPageChange={(e, { activePage }) => setActivePage(activePage)}
                    size='mini'
                    activePage={activePage}
                    totalPages={Math.ceil(unitsLength / 10)} />
            </Grid.Row>
        </Grid>

    </>)
}

export default HomePage