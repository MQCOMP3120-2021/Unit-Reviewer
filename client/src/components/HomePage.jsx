import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import {
  Grid, Card, Rating, Pagination,
} from 'semantic-ui-react';
import Loading from './Loading';

const HomePage = ({ appLoaded, unitsLength, getUnits }) => {
  const [actPage, setActPage] = useState(1);
  const [homeUnits, setHomeUnits] = useState([]);
  const [loaded, setLoaded] = useState(appLoaded);
  const [loading, setLoading] = useState();
  console.log(Math.ceil(unitsLength / 10));

  const getHomeUnits = async () => {
    console.log('hi');
    console.log('active Page: ', actPage);
    let info = [];
    if (loaded.length < 1) {
      console.log('new load');
      setLoading(true);
      info = await getUnits(actPage, true);
    } else {
      setLoading(true);
      info = await getUnits(actPage);
    }
    // console.log("info: ",info)
    const allUnits = info[0];
    const pages = info[1];
    // console.log('loaded: ', pages);
    // console.log('allUnits: ', allUnits);
    // console.log("allUnits last: ",allUnits[allUnits.length-1])
    setLoaded(pages);
    setHomeUnits(allUnits);
    setLoading(false);
  };

  useEffect(() => {
    console.log('active page is: ', actPage);
    setHomeUnits([]);
    getHomeUnits();
  }, [actPage]);

  return (
    <>
      <Grid padded centered>
        <Grid.Row>
          {loading ? (
            <>
              {[...Array(10).keys()].map(() => (
                <div>
                  <Loading />
                </div>
              ))}
            </>
          ) : (
            homeUnits
              .filter((elem, idx) => {
                const c = loaded.indexOf(actPage) * 10;
                return idx >= c && idx < c + 10;
              })
              .map((item) => (
                <Link key={item._id} to={`unit/${item._id}`}>
                  <Card
                    style={{ marginBottom: 10, marginTop: 10, marginRight: 10 }}
                  >
                    <Card.Content header={item.code} />
                    <Card.Content description={item.title} />
                    <Card.Content extra>
                      <Rating
                        icon="star"
                        defaultRating={
                          item.reviews.length > 0
                          && item.reviews
                            .map((rev) => rev.rating)
                            .reduce((a, b) => a + b) / item.reviews.length
                        }
                        disabled
                        maxRating={5}
                      />
                      {' '}
                      (
                      {item.reviews.length === 1
                        ? `${item.reviews.length} Rating`
                        : `${item.reviews.length} Ratings`}
                      )
                    </Card.Content>
                  </Card>
                </Link>
              ))
          )}
        </Grid.Row>
        <Grid.Row>
          <Pagination
            onPageChange={(e, { activePage }) => {
              setActPage(activePage);
            }}
            size="mini"
            activePage={actPage}
            totalPages={Math.ceil(unitsLength / 10)}
          />
        </Grid.Row>
      </Grid>
    </>
  );
};

export default HomePage;
