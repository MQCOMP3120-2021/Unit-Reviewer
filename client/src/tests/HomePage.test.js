/**
 * @jest-environment jsdom
 */
import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import { fireEvent, render } from '@testing-library/react'
import fs from 'fs'

import { BrowserRouter as Router } from 'react-router-dom'
import HomePage from '../HomePage'

const sampleData = (fileName) => {
    const rawData = fs.readFileSync(fileName)
    const data = JSON.parse(rawData)
    return data
}

describe("Homepage Display", () => {
    test('hompage displays all units', () => {
        const units = sampleData('src/tests/samples/units.json')
        const component = render(
            <Router>
                <HomePage units={units} />
            </Router>
        )

        units.map(u => expect(component.container).toHaveTextContent(u.title))
    })

    test('snapshot test', () => {
        const units = sampleData('src/tests/samples/units.json')

        const component = render(
            <Router>
                <HomePage units={units} />
            </Router>
        )

        expect(component).toMatchSnapshot()
    })
})