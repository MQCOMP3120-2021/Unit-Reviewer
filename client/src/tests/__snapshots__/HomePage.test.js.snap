// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage Display snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui centered padded grid"
      >
        <a
          href="/unit/615f9936490575598b5a0835"
        >
          <div
            class="ui card"
            style="margin-bottom: 10px; margin-top: 10px; margin-right: 10px;"
          >
            <div
              class="content"
            >
              <div
                class="header"
              >
                COMP2320
              </div>
            </div>
            <div
              class="content"
            >
              <div
                class="description"
              >
                Offensive Security
              </div>
            </div>
            <div
              class="extra content"
            >
              <div
                class="ui star disabled rating"
                role="radiogroup"
                tabindex="0"
              >
                <i
                  aria-checked="false"
                  aria-posinset="1"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="2"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="3"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="4"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="5"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
              </div>
               (
              6
               Ratings)
            </div>
          </div>
        </a>
        <a
          href="/unit/615f9a3f490575598b5a0854"
        >
          <div
            class="ui card"
            style="margin-bottom: 10px; margin-top: 10px; margin-right: 10px;"
          >
            <div
              class="content"
            >
              <div
                class="header"
              >
                COMP5323
              </div>
            </div>
            <div
              class="content"
            >
              <div
                class="description"
              >
                345634534t45h45htr4
              </div>
            </div>
            <div
              class="extra content"
            >
              <div
                class="ui star disabled rating"
                role="radiogroup"
                tabindex="0"
              >
                <i
                  aria-checked="false"
                  aria-posinset="1"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="2"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="true"
                  aria-posinset="3"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="4"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="5"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
              </div>
               (
              4
               Ratings)
            </div>
          </div>
        </a>
        <a
          href="/unit/6163d3bfdd4d7c1df98aa57e"
        >
          <div
            class="ui card"
            style="margin-bottom: 10px; margin-top: 10px; margin-right: 10px;"
          >
            <div
              class="content"
            >
              <div
                class="header"
              >
                COMP34634
              </div>
            </div>
            <div
              class="content"
            >
              <div
                class="description"
              >
                ntrgmnhmyytju,,,,,u.
              </div>
            </div>
            <div
              class="extra content"
            >
              <div
                class="ui star disabled rating"
                role="radiogroup"
                tabindex="0"
              >
                <i
                  aria-checked="false"
                  aria-posinset="1"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="true"
                  aria-posinset="2"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="3"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="4"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="5"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
              </div>
               (
              2
               Ratings)
            </div>
          </div>
        </a>
        <a
          href="/unit/6167684c2f84f347c78fece9"
        >
          <div
            class="ui card"
            style="margin-bottom: 10px; margin-top: 10px; margin-right: 10px;"
          >
            <div
              class="content"
            >
              <div
                class="header"
              >
                ENGG3465
              </div>
            </div>
            <div
              class="content"
            >
              <div
                class="description"
              >
                tgrtjy
              </div>
            </div>
            <div
              class="extra content"
            >
              <div
                class="ui star disabled rating"
                role="radiogroup"
                tabindex="0"
              >
                <i
                  aria-checked="false"
                  aria-posinset="1"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="2"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="3"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="4"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="5"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
              </div>
               (
              2
               Ratings)
            </div>
          </div>
        </a>
        <a
          href="/unit/61676be72f84f347c78fed15"
        >
          <div
            class="ui card"
            style="margin-bottom: 10px; margin-top: 10px; margin-right: 10px;"
          >
            <div
              class="content"
            >
              <div
                class="header"
              >
                PHYS45645
              </div>
            </div>
            <div
              class="content"
            >
              <div
                class="description"
              >
                trhrtjhyktyuyikty
              </div>
            </div>
            <div
              class="extra content"
            >
              <div
                class="ui star disabled rating"
                role="radiogroup"
                tabindex="0"
              >
                <i
                  aria-checked="false"
                  aria-posinset="1"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="2"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="3"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="4"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="5"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
              </div>
               (
              3
               Ratings)
            </div>
          </div>
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui centered padded grid"
    >
      <a
        href="/unit/615f9936490575598b5a0835"
      >
        <div
          class="ui card"
          style="margin-bottom: 10px; margin-top: 10px; margin-right: 10px;"
        >
          <div
            class="content"
          >
            <div
              class="header"
            >
              COMP2320
            </div>
          </div>
          <div
            class="content"
          >
            <div
              class="description"
            >
              Offensive Security
            </div>
          </div>
          <div
            class="extra content"
          >
            <div
              class="ui star disabled rating"
              role="radiogroup"
              tabindex="0"
            >
              <i
                aria-checked="false"
                aria-posinset="1"
                aria-setsize="5"
                class="active icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="2"
                aria-setsize="5"
                class="active icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="3"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="4"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="5"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
            </div>
             (
            6
             Ratings)
          </div>
        </div>
      </a>
      <a
        href="/unit/615f9a3f490575598b5a0854"
      >
        <div
          class="ui card"
          style="margin-bottom: 10px; margin-top: 10px; margin-right: 10px;"
        >
          <div
            class="content"
          >
            <div
              class="header"
            >
              COMP5323
            </div>
          </div>
          <div
            class="content"
          >
            <div
              class="description"
            >
              345634534t45h45htr4
            </div>
          </div>
          <div
            class="extra content"
          >
            <div
              class="ui star disabled rating"
              role="radiogroup"
              tabindex="0"
            >
              <i
                aria-checked="false"
                aria-posinset="1"
                aria-setsize="5"
                class="active icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="2"
                aria-setsize="5"
                class="active icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="true"
                aria-posinset="3"
                aria-setsize="5"
                class="active icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="4"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="5"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
            </div>
             (
            4
             Ratings)
          </div>
        </div>
      </a>
      <a
        href="/unit/6163d3bfdd4d7c1df98aa57e"
      >
        <div
          class="ui card"
          style="margin-bottom: 10px; margin-top: 10px; margin-right: 10px;"
        >
          <div
            class="content"
          >
            <div
              class="header"
            >
              COMP34634
            </div>
          </div>
          <div
            class="content"
          >
            <div
              class="description"
            >
              ntrgmnhmyytju,,,,,u.
            </div>
          </div>
          <div
            class="extra content"
          >
            <div
              class="ui star disabled rating"
              role="radiogroup"
              tabindex="0"
            >
              <i
                aria-checked="false"
                aria-posinset="1"
                aria-setsize="5"
                class="active icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="true"
                aria-posinset="2"
                aria-setsize="5"
                class="active icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="3"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="4"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="5"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
            </div>
             (
            2
             Ratings)
          </div>
        </div>
      </a>
      <a
        href="/unit/6167684c2f84f347c78fece9"
      >
        <div
          class="ui card"
          style="margin-bottom: 10px; margin-top: 10px; margin-right: 10px;"
        >
          <div
            class="content"
          >
            <div
              class="header"
            >
              ENGG3465
            </div>
          </div>
          <div
            class="content"
          >
            <div
              class="description"
            >
              tgrtjy
            </div>
          </div>
          <div
            class="extra content"
          >
            <div
              class="ui star disabled rating"
              role="radiogroup"
              tabindex="0"
            >
              <i
                aria-checked="false"
                aria-posinset="1"
                aria-setsize="5"
                class="active icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="2"
                aria-setsize="5"
                class="active icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="3"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="4"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="5"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
            </div>
             (
            2
             Ratings)
          </div>
        </div>
      </a>
      <a
        href="/unit/61676be72f84f347c78fed15"
      >
        <div
          class="ui card"
          style="margin-bottom: 10px; margin-top: 10px; margin-right: 10px;"
        >
          <div
            class="content"
          >
            <div
              class="header"
            >
              PHYS45645
            </div>
          </div>
          <div
            class="content"
          >
            <div
              class="description"
            >
              trhrtjhyktyuyikty
            </div>
          </div>
          <div
            class="extra content"
          >
            <div
              class="ui star disabled rating"
              role="radiogroup"
              tabindex="0"
            >
              <i
                aria-checked="false"
                aria-posinset="1"
                aria-setsize="5"
                class="active icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="2"
                aria-setsize="5"
                class="active icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="3"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="4"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
              <i
                aria-checked="false"
                aria-posinset="5"
                aria-setsize="5"
                class="icon"
                role="radio"
                tabindex="-1"
              />
            </div>
             (
            3
             Ratings)
          </div>
        </div>
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
